Refactor API for unconventionally-indexed arrays

timholy opened the pull request:
This is effectively "cleanup" from #16260. It gets rid of `shape` and renames `allocate_for` to be `similar`. It achieves these by introducing a new Range type, `OneTo(n)`, which is equivalent (in a value-sense) to `1:n`. These are changes recommended by @JeffBezanson. I'm hoping to avoid having `OneTo` "leak" out into user space, so it's not exported, but I suspect we may need to export it at some point (let's first see how far we can get without doing that).

Despite the name of this branch, it does not make unconventional indices safer (ref `@arraysafe` in #16973); it seems better to do that in a separate PR.

A key thing here is going to be to find out whether I've done something nasty for performance, so @nanosoldier `runbenchmarks(ALL, vs=:master)`.

timholy committed:
Add specialized Range type, OneTo, for use with indices

timholy committed:
Update broadcast and its callers to the new indices behavior

timholy committed:
Updates for subarray for new indices

pabloferz commented on base/abstractarray.jl:
```diff
@@ -348,59 +326,54 @@ different element type it will create a regular `Array` instead:
-_similar(::IndicesBehavior, a::AbstractArray, T::Type)   = similar(a, T, indices(a))
+to_shape(::Tuple{}) = ()
+to_shape(dims::Dims) = dims
+to_shape(dims::DimsOrInds) = map(to_shape, dims)
```
Another idea (I don't know how convenient or cumbersome it'd be) would be to have a type `Shape` of `Shp`. This way this could changed to `convert` methods so one can write `Shape(dims)` and `convert(Shape, dim)`.

timholy replied:
`to_shape` is basically a compatibility call, packaging `AbstractUnitRange` objects as `Dims`-tuples when possible (when the range can be guaranteed, via the type system, to start at 1). In that sense we already have the type(alias).

timholy committed:
Rename allocate_for -> similar. Fixes #17124.

timholy committed:
Iteration improvements for arrays with non-1 indices

Better to have this logic in the main functions rather than in specific packages

timholy committed:
Replace shape with indices, and redesign similar

timholy commented:
I seem to have messed up the nanosoldier call, so let's try again after adding quotes: @nanosoldier `runbenchmarks(ALL, vs=":master")`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17137#issuecomment-228703462) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/75370acde884780e55a4b669a4ac2dc0e6bd45a8/b1b77a1_vs_a543e0c/report.md). cc @jrevels

timholy committed:
Eliminate the indicesbehavior trait

timholy committed:
Performance improvements for indices

timholy commented:
I've canceled the CI here; while the runtime looks good, 1b684d5 ("eliminate the `indicesbehavior` trait") absolutely hammers compile time.

Master:
```jl
julia> using Base.Test

shell> cd test/linalg/
/home/tim/src/julia-0.5b/test/linalg

julia> @time include("schur.jl")
 17.479364 seconds (13.13 M allocations: 516.591 MB, 1.47% gc time)

julia> @time include("schur.jl")
  1.011176 seconds (445.60 k allocations: 23.307 MB, 1.33% gc time)
```

This PR:
```jl
julia> using Base.Test

shell> cd test/linalg/
/home/tim/src/julia-0.5/test/linalg

julia> @time include("schur.jl")
 44.836752 seconds (72.31 M allocations: 3.040 GB, 2.34% gc time)

julia> @time include("schur.jl")
  0.902519 seconds (447.55 k allocations: 23.480 MB, 1.02% gc time)
```
I'll see what I can figure out. Meanwhile, it's a nice test case for folks who like to look at compiler performance :wink: .

timholy commented:
Wow, the ProfileView looks like cathedrals of colorful `inference` sand, reaching towards the heavens. So pretty and spiritual, I think I'll frame it.
![image](https://cloud.githubusercontent.com/assets/1525481/16467802/97208b60-3e0e-11e6-952c-af3b4e094207.png)

timholy commented:
Interesting observation: if I comment out [these](https://github.com/JuliaLang/julia/blob/70a7da94a4e3a28819e61992b929f09f5a1f363d/test/linalg/schur.jl#L19) [loops](https://github.com/JuliaLang/julia/blob/70a7da94a4e3a28819e61992b929f09f5a1f363d/test/linalg/schur.jl#L23) and just manually set the values, then master and this PR have similar compile times. But set `eltya = Int` and leave the `atype` loop intact, and there's a nearly 3x difference in compile times. Also possibly-relevant, changing the `for` loop to a `while` loop,
```diff
diff --git a/test/linalg/schur.jl b/test/linalg/schur.jl
index 5370afd..02a07dc 100644
--- a/test/linalg/schur.jl
+++ b/test/linalg/schur.jl
@@ -16,11 +16,17 @@ srand(1234321)
 areal = randn(n,n)/2
 aimg  = randn(n,n)/2
 
-for eltya in (Float32, Float64, Complex64, Complex128, Int)
+# for eltya in (Float32, Float64, Complex64, Complex128, Int)
+eltya = Int
     a = eltya == Int ? rand(1:7, n, n) : convert(Matrix{eltya}, eltya <: Complex ? complex(areal, aimg) : areal)
     asym = a'+a                  # symmetric indefinite
     apd  = a'*a                 # symmetric positive-definite
-    for atype in ("Array", "SubArray")
+    indx = 1
+    atypes = ["Array", "SubArray"]
+    while indx <= length(atypes)
+#    for atype in ("Array", "SubArray")
+        atype = atypes[indx]
+        indx += 1
         if atype == "Array"
             a = a
         else
@@ -96,4 +102,4 @@ for eltya in (Float32, Float64, Complex64, Complex128, Int)
         @test NS[:S] ≈ sS
         @test NS[:Z] ≈ sZ
     end
-end
+# end
```
does **not** fix the performance, so this seems different from #16122.

timholy commented:
Here's a minimal reproducer:
```jl
using Base.Test

a = rand(1:7, 10, 10)
for atype in ("Array", "SubArray")
    d,v = eig(a)
end
```
Master:
```jl
julia> @time include("/tmp/tim/schur.jl")
  1.740690 seconds (1.93 M allocations: 84.396 MB, 2.91% gc time)
```
This PR:
```jl
julia> @time include("/tmp/tim/schur.jl")
 13.267818 seconds (46.09 M allocations: 1.985 GB, 5.30% gc time)
```
But now comment out the `for` loop, and they are the same.

JeffBezanson commented:
Wow, interesting. What happens if you use an array instead of a tuple: `for atype in ["Array", "SubArray"]`?

timholy commented:
> Wow, interesting.

In the Chinese curse-sense, yes.  :smile:

> What happens if you use an array instead of a tuple: for atype in ["Array", "SubArray"]?

Still slow.

I've debugged this a little further; one relevant point is that I can call 
```
a = rand(1:7, 10, 10)
eig(a)
```
before `include`ing the file with the loop (or copy/pasting into the REPL), and it's *still* dirt-slow. But I can't take the `eig` call out of the loop without making it fast. Finally, it's all one giant toplevel-inference call: this diff
```diff
+
+type InfRef
+    val::Bool
+end
+const debug = InfRef(false)
 function typeinf_ext(linfo::LambdaInfo)
     if isdefined(linfo, :def)
+        local tstart
+        if debug.val
+            print(linfo.def.name, " start: ", ccall(:jl_clock_now, Float64, ()))
+        end
         # method lambda - infer this specialization via the method cache
         (code, _t, inferred) = typeinf_edge(linfo.def, linfo.specTypes, linfo.sparam_vals, true, true, true, linfo)
         if inferred && code.inferred && linfo !== code
@@ -1567,13 +1576,22 @@ function typeinf_ext(linfo::LambdaInfo)
             linfo.inferred = true
             linfo.inInference = false
         end
+        if debug.val
+            print(linfo.def.name, " stop: ", ccall(:jl_clock_now, Float64, ()))
+        end
         return code
     else
         # toplevel lambda - infer directly
+        if debug.val
+            println("toplevel inference")
+        end
         linfo.inInference = true
         frame = InferenceState(linfo, true, true)
         typeinf_loop(frame)
         @assert frame.inferred # TODO: deal with this better
+        if debug.val
+            println("toplevel inference done")
+        end
         return linfo
     end
 end
```
gives this output (having already called `eig` once):
```
julia> for atype in ("Array", "SubArray")
           d,v = eig(a)
       end
toplevel inference
toplevel inference done
:put! start: 1.46724e+09:put! stop: 1.46724e+09:start start: 1.46724e+09:start stop: 1.46724e+09:indexed_next start: 1.46724e+09:indexed_next stop: 1.46724e+09
```
and all the waiting is before the "toplevel inference done".

StirlingNewberry commented:
宁為太平犬，莫做亂离人

Though the actual curse is English.

On Wed, Jun 29, 2016 at 6:10 PM, Tim Holy <notifications@github.com> wrote:

> Wow, interesting.
>
> In the Chinese curse-sense, yes. 😄
>
> What happens if you use an array instead of a tuple: for atype in
> ["Array", "SubArray"]?
>
> Still slow.
>
> I've debugged this a little further; one relevant point is that I can call
>
> a = rand(1:7, 10, 10)
> eig(a)
>
> before includeing the file with the loop (or copy/pasting into the REPL),
> and it's *still* dirt-slow. But I can't take the eig call out of the loop
> without making it fast. Finally, it's all one giant toplevel-inference
> call: this diff
>
> ++type InfRef+    val::Bool+end+const debug = InfRef(false)
>  function typeinf_ext(linfo::LambdaInfo)
>      if isdefined(linfo, :def)+        local tstart+        if debug.val+            print(linfo.def.name, " start: ", ccall(:jl_clock_now, Float64, ()))+        end
>          # method lambda - infer this specialization via the method cache
>          (code, _t, inferred) = typeinf_edge(linfo.def, linfo.specTypes, linfo.sparam_vals, true, true, true, linfo)
>          if inferred && code.inferred && linfo !== code@@ -1567,13 +1576,22 @@ function typeinf_ext(linfo::LambdaInfo)
>              linfo.inferred = true
>              linfo.inInference = false
>          end+        if debug.val+            print(linfo.def.name, " stop: ", ccall(:jl_clock_now, Float64, ()))+        end
>          return code
>      else
>          # toplevel lambda - infer directly+        if debug.val+            println("toplevel inference")+        end
>          linfo.inInference = true
>          frame = InferenceState(linfo, true, true)
>          typeinf_loop(frame)
>          @assert frame.inferred # TODO: deal with this better+        if debug.val+            println("toplevel inference done")+        end
>          return linfo
>      end
>  end
>
> gives this output (having already called eig once):
>
> julia> for atype in ("Array", "SubArray")
>            d,v = eig(a)
>        end
> toplevel inference
> toplevel inference done
> :put! start: 1.46724e+09:put! stop: 1.46724e+09:start start: 1.46724e+09:start stop: 1.46724e+09:indexed_next start: 1.46724e+09:indexed_next stop: 1.46724e+09
>
> and all the waiting is before the "toplevel inference done".
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/17137#issuecomment-229504477>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ATKyInTNYpCYEmrmPTZtDE8iCzFUBMtXks5qQu1FgaJpZM4I-tDS>
> .
>

JeffBezanson commented:
Inference seems to be encountering a large number of different signatures for `_indices`, all similar to

```
Tuple{Base.#_indices, Tuple{Base.OneTo{Int64}, Base.OneTo, Base.OneTo, Base.OneTo, Base.OneTo, Base.OneTo, Base.OneTo, Base.OneTo, Base.OneTo}, Int64, Base.SubArray{Base.Complex{Float32}, 2, A<:Union{Base.ReshapedArray{T<:Any, N<:Any, A<:DenseArray, MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}, N<:Any}}}, DenseArray}, Tuple{Union{Base.UnitRange{Int64}, Base.Colon}, Vararg{Union{Base.AbstractCartesianIndex, Base.Range{Int64}, Int64, Base.Colon}, N<:Any}}, true}, Union{Base.AbstractCartesianIndex{N<:Any}, Base.Range{Int64}, Int64, Base.Colon}, Vararg{Union{Base.AbstractCartesianIndex{N<:Any}, Base.Range{Int64}, Int64, Base.Colon}, N<:Any}}
```

timholy commented:
Surprising, because there are exactly [6](https://github.com/JuliaLang/julia/blob/e1ef74ade1f01ce54e7d0d028914da3edf919e91/base/abstractarray.jl#L46-L50) [definitions](https://github.com/JuliaLang/julia/blob/e1ef74ade1f01ce54e7d0d028914da3edf919e91/base/subarray.jl#L283-L286) in this PR, fewer than the 8 there are in current Base. But of course they're called differently. The example you showed suggests something is calling it with 9 indices...given that we're talking about matrix algebra (2 dimensions), what the heck is that from?

JeffBezanson commented:
It's from recursion in the inference process itself.

timholy committed:
subarray indices: work around an inference hang (#17278)

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`.

timholy committed:
Fix offsetarray tests and move them earlier

The reason to move earlier is to test whether the methods corrupt
other operations.

timholy committed:
Simpler sub2ind/ind2sub for Tuple{OneTo}

timholy commented:
@jrevels, OK to start nanosoldier now? (It's this PR I'm most interested in right now.)

jrevels commented:
The [daily build ran successfully](https://github.com/JuliaLang/julia/commit/fa5af2377381145c8949c2de7c2bfceaf772d83e#commitcomment-18142819), so I think it's good now:

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17137#issuecomment-230835735) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/050e7218b61bb17c623599ba4522a6f686261361/362f5e8_vs_853317d/report.md). cc @jrevels

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17137#issuecomment-230965839) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/55f76c534e723fd225bb14bab61bf86e9a19c326/50c986e_vs_a8bca12/report.md). cc @jrevels

timholy committed:
Elim. `indicesperformance` and use only a single path for bounds-checking

timholy committed:
Reduce use of splatting in sparse matrix size operations

Splatting forces dynamic method lookup in places where that's a major cost

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17137#issuecomment-231056378) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6d4002a546bfb95dffb070271a69f60de99a73a3/cce8fa6_vs_36d36b0/report.md). cc @jrevels

timholy committed:
Fixes for indices1(::Subarray) and indices1 specializations

This leads to noticeable performance improvements for several benchmarks

timholy commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`. My fingers type that almost automatically now.

tkelman commented on base/abstractarray.jl:
```diff
@@ -348,79 +263,54 @@ different element type it will create a regular `Array` instead:
-of arrays that may have unconventional indexing (starting at other
-than 1), as determined by `referencearray` and the optional `shape`
-information.
+`storagetype`, but with `indices` specified by the last
```
needs genstdlib run?

timholy replied:
done

tkelman commented on base/range.jl:
```diff
@@ -355,20 +364,21 @@ let smallint = (Int === Int64 ?
 
-last{T}(r::StepRange{T}) = r.stop
-last(r::UnitRange) = r.stop
+last(r::OrdinalRange) = r.stop
```
are we sure all `OrdinalRange` types have a `stop` field?

timholy replied:
Not guaranteed, but I don't think there's a better fallback definition. (It also happens to work for all `OrdinalRange`s currently defined in `Base`.) Anyone can of course specialize

tkelman commented on base/range.jl:
```diff
@@ -487,6 +511,7 @@ end
 
 show(io::IO, r::Range) = print(io, repr(first(r)), ':', repr(step(r)), ':', repr(last(r)))
 show(io::IO, r::UnitRange) = print(io, repr(first(r)), ':', repr(last(r)))
+show(io::IO, r::OneTo) = print(io, "OneTo(", r.stop, ")")
```
if this isn't exported, should it be shown qualified as `Base.OneTo` ?

timholy replied:
Yes, good catch.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17137#issuecomment-231158603) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/487409f1a825b6ab5f6b22fdc509a9ddfa38f7fb/e8efdfd_vs_3704eac/report.md). cc @jrevels

timholy committed:
Update the docs

timholy commented:
I have a good feeling (based on local benchmarking) about this one: @nanosoldier `runbenchmarks(ALL, vs = ":master")`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/17137#issuecomment-231215904) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/454c2b55b32e8eab4a8d0a4b46076d513c775af0/c0761cc_vs_e90bcf3/report.md). cc @jrevels

timholy commented:
Fewer exports, cleaner API, fewer lines of code, and more performant---what's not to like? Advance warning, I'll be merging as soon as CI finishes---I don't have any reason to suspect that the one test failure (so far) has anything to do with this PR.

tkelman commented:
It looks like this broke several packages with a `MethodError: no method matching DataArrays.DataArray{T,N}(::Type{Bool}, ::Tuple{Base.OneTo{Int64}})`, see PkgEval results for DataArrays and others

tkelman commented:
Looks like the issue in DataArrays stems from it using `broadcast_shape`. That now returns `OneTo` objects that packages don't know what to do with. Tying into the unexported broadcast internals isn't ideal, but I suspect DataArrays is not the only package that may have needed to do that. I'm leaning in the direction of walking back on trying to support unconventionally indexed arrays for 0.5, since we really would like to have a package ecosystem that at least half works as it does on 0.4 by the time we have the final 0.5.0 done.

timholy commented:
If they need a quick-fix, they should be able to just call `Base.to_shape` on the output of `broadcast_shape`. That said, just waiting for some consensus might be the better option.

tkelman commented:
I'm stuck at https://github.com/JuliaStats/DataArrays.jl/pull/205. Help appreciated.
